File: src/main/java/thirtytwo/degrees/AppConfig.java
package thirtytwo.degrees;

import com.netflix.config.*;
import org.springframework.context.annotation.*;
import org.springframework.context.annotation.ComponentScan.Filter;
import org.springframework.stereotype.Controller;

import javax.inject.Named;

/**
 * User: gibbsb
 * Date: 9/21/12
 * Time: 4:18 PM
 * <p/>
 * http://static.springsource.org/spring/docs/3.1.x/spring-framework-reference/html/mvc.html#mvc-config-enable
 */
@Configuration
@ComponentScan(basePackageClasses = AppConfig.class, excludeFilters = {
    @Filter(Controller.class)
})
public class AppConfig {

    //TODO: create a proxy that auto-populates a pojo full off values and watches for changes?
    // https://github.com/Netflix/archaius/wiki/Getting-Started
    @Bean @Named("helloText")
    public DynamicStringProperty helloText() {
        return DynamicPropertyFactory.getInstance().getStringProperty("hello.text", "Hello default text");
    }
}

File: scala_2.9.1/src/main/scala/thirtytwo/degrees/halfpipe/scalaexample/AppConfig.scala
package thirtytwo.degrees.halfpipe.scalaexample

import org.springframework.context.annotation._
import com.netflix.config.scala.DynamicProperties
import javax.inject.Named
import org.springframework.context.annotation.ComponentScan.Filter
import org.springframework.stereotype.Controller
import com.codahale.jersey.inject.ScalaCollectionsQueryParamInjectableProvider
import com.codahale.jerkson.ScalaModule
import com.fasterxml.jackson.datatype.guava.GuavaModule
import thirtytwo.degrees.halfpipe.jersey.HalfpipeScalaObjectMapperProvider
import com.yammer.metrics.util.DeadlockHealthCheck
import scala.collection.JavaConversions._
import com.yammer.metrics.core.HealthCheck
import com.yammer.metrics.HealthChecks
import java.util.{List => JList}

@Configuration
@ComponentScan (basePackageClasses = Array (classOf[AppConfig]),
  excludeFilters = Array(new Filter (Array (classOf[Controller], classOf[Configuration]))))
@ImportResource(Array ("classpath:META-INF/spring/applicationContext-security.xml"))
class AppConfig extends DynamicProperties {

  @Bean @Named("helloText")
  def helloText() = dynamicStringProperty("hello.text", "Hello default")

  @Bean
  def deadlockHealthCheck() = new DeadlockHealthCheck()

  @Bean
  def healthChecks(healthChecks: JList[HealthCheck]) = {
    healthChecks foreach { healthCheck =>
      HealthChecks.register(healthCheck)
    }
    HealthChecks.defaultRegistry()
  }

  @Bean @Scope
  def jacksonScalaCollections() = new ScalaCollectionsQueryParamInjectableProvider()

  @Bean
  def scalaModule() = new ScalaModule(Thread.currentThread().getContextClassLoader)

  @Bean
  def guavaModule() = new GuavaModule

  @Bean @Scope
  def objectMapperProvider() = new HalfpipeScalaObjectMapperProvider(scalaModule(), guavaModule())

}